File: cs50x/pset6/sentimental-readability/readability.py
from cs50 import get_string

# function that returns grade level based on cli


def get_grade_level(cli):
    if cli in range(17):
        return ("Grade " + str(cli))
    elif cli > 16:
        return ("Grade 16+")
    else:
        return ("Before Grade 1")


def main():
    text = get_string("Text: ")
    # split text into list of words
    word_list = text.split()
    sentences_num = 0
    letter_num = 0

    # check each word
    for word in word_list:
        # if last char in word is ., !, or ? increment number of sentences
        if word[-1] in ['.', '!', '?']:
            sentences_num += 1
        # if last char is ' or ", and second to last is ., !, ? increment number of sentences
        if word[-1] in ['\'', '"']:
            if word[-2] in ['.', '!', '?']:
                sentences_num += 1
        # for each letter in word check if it's alphanumerical and increment number of letters
        for letter in word:
            if (letter.isalnum()):
                letter_num += 1
    # calculate average number of letters per 100 words in the text
    cli_l = (letter_num / len(word_list)) * 100.00
    # calculate average number of sentences per 100 words in the text
    cli_s = (sentences_num / len(word_list)) * 100.00
    # calculate coleman-liau index
    cli = round((0.0588 * cli_l - 0.296 * cli_s - 15.8))
    # print grade level
    print(get_grade_level(cli))


main()

File: cs50x/pset2/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int endOfSentence(char c);
void getCounters(double ctrs[], string text);
void gradeLevel(double lvl);
int main(void)
{
    string text = get_string("Text: ");
    // Array that represents, in order, number of letters, number of words, and number of sentences.
    double counters[] = {0.0, 0.0, 0.0};
    double L, S, CLI;
    getCounters(counters, text);
    L = ((counters[0]) / (counters[1])) * 100.0;
    S = ((counters[2]) / (counters[1])) * 100.0;
    CLI = round((0.0588 * L) - (0.296 * S) - 15.8);
    gradeLevel(CLI);
}

int endOfSentence(char c)
{
    if (((c == '.') || (c == '!') || (c == '?')))
    {
        return 1;
    }
    return 0;
}
void getCounters(double ctrs[], string text)
{
    for (int i = 0, len = strlen(text); i < len; i++)
    {
        if (isalnum(text[i]))
        {
            ctrs[0]++;
        }
        else if (endOfSentence(text[i]))
        {
            if (i + 1 == len)
            {
                ctrs[1]++;
            }
            if (endOfSentence(text[i - 1]))
            {
                continue;
            }
            ctrs[2]++;
        }
        else if (isspace(text[i]))
        {
            if (isspace(text[i - 1]))
            {
                continue;
            }
            ctrs[1]++;
        }
    }
}

void gradeLevel(double lvl)
{
    int readingLvl = (int) lvl;
    if (readingLvl > 16)
    {
        readingLvl = 20;
    }
    printf("%i\n", readingLvl);
    switch ((int) readingLvl)
    {
        case 20:
            printf("Grade 16+\n");
            break;
        case 1:
        case 2:
        case 3:
        case 4:
        case 5:
        case 6:
        case 7:
        case 8:
        case 9:
        case 10:
        case 11:
        case 12:
        case 13:
        case 14:
        case 15:
        case 16:
            printf("Grade %i\n", (int) lvl);
            break;
        default:
            printf("Before Grade 1\n");
            break;
    }
}
